docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,389 @@
+
+#### Story 4
+
+As an IPAM provider I want a single interface to integrate with all infrastructure providers that can benefit from IPAM integration.
+
+### Implementation Details/Notes/Constraints
+
+#### New API Types
+
+The following API Types should be added to cluster-api. In the first iteration they will be added to the experimental API and will therefore live in the `cluster.x-k8s.io` group.
+
+##### IPAddressClaim
+
+The IPAddressClaim is used to request ip addresses from a pool. It gets reconciled by IPAM providers, which can filter based on the `Spec.Pool` reference.
+
+After an IPAddressClaim is created, the `Spec.Pool` reference, and therefore essentially the entire `Spec` is immutable to simplify reconciliation of the resource.
+
+```go
+// IPAddressClaimSpec describes the desired state of an IPAddressClaim
+type IPAddressClaimSpec struct {
+  // Pool is a reference to the pool from which an IP address should be allocated.
+  Pool LocalObjectReference `json:"pool,omitempty"`
+}
+
+// IPAddressClaimStatus contains the status of an IPAddressClaim
+type IPAddressClaimStatus struct {
+  // Address is a reference to the address that was allocated for this claim.
+  Address LocalObjectReference `json:"address,omitempty"`
+
+  // Conditions provide details about the status of the claim.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+ 
+// IPAddressClaim can be used to allocate IPAddresses from an IP Pool.
+type IPAddressClaim struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+  
+  Spec   IPAddressClaimSpec   `json:"spec,omitempty"`
+  Status IPAddressClaimStatus `json:"status,omitempty"`
+}
+```
+
+##### IPAddress
+
+IPAddress resources are created to fulfil an IPAddressClaim. They are created by the IPAM provider that reconciles the claim.
+
+IPAddresses are immutable.
+
+```go
+// IPAddressSpec describes an IPAddress
+type IPAddressSpec struct {
+  // Claim is a reference to the claim this IPAddress was created for.
+  Claim LocalObjectReference `json:"claim,omitempty"`
+ 
+  // Pool is a reference to the pool that this IPAddress was created from.
+  Pool LocalObjectReference `json:"pool,omitempty"`
+ 
+  // Address is the IP address.
+  Address string `json:"address"`
+ 
+  // Prefix is the prefix of the address.
+  Prefix int `json:"prefix,omitempty"`
+ 
+  // Gateway is the network gateway of network the address is from.
+  Gateway string `json:"gateway,omitempty"`
+}
+ 
+// IPAddress is a representation of an IP Address that was allocated from an IP Pool.
+type IPAddress struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec IPAddressSpec `json:"spec,omitempty"`
+}
+```
+
+##### New Reference Type
+
+```go
+type LocalObjectReference struct {
+  Group string
+  Kind string
+  Name string
+}
+```
+
+#### Implementing an IPAM Provider
```

Comment:
I think a sequence diagram how different components work together to create an IPAddress could be helpful

Reply:
Too much freestyle or is this sufficient?
![image](https://user-images.githubusercontent.com/1710904/157898609-dbb556a4-c89b-48ca-96aa-c210be013ca5.png)